{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d5de1406-c6a3-4df5-96de-85c6a1037c92",
   "metadata": {},
   "source": [
    "# Finding the Total Number of Bases in a Gene List"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "315d0062-3713-4106-8998-f5ec742f5fa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### standard imports\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "import plotly.express as px\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "d638cb79-77e9-4617-9c1b-c061c04985d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Exons (codons) tested</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AKT1 (NM_005163)</td>\n",
       "      <td>3 (16-56), 4 (59-96), 6 (146-149), 6-7 (177-19...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AKT2 (NM_001626)</td>\n",
       "      <td>3 (16-55), 4 (76-96), 6 (148-153), 7 (192-210)...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AKT3 (NM_005465)</td>\n",
       "      <td>2 (16-28), 2 (48-58), 3 (66-95), 4 (123-143), ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ALK (NM_004304)</td>\n",
       "      <td>20-28 (1096-1388)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AR (NM_000044)</td>\n",
       "      <td>1 (1-45), 1 (135-178), 1 (245-285), 1 (342-388...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>TP53 (NM_000546)</td>\n",
       "      <td>2-11 (1-394)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>TSC1 (NM_000368)</td>\n",
       "      <td>3-23 (1-1165)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>TSC2 (NM_000548)</td>\n",
       "      <td>2-35 (1-1523), 37-42 (1555-1808)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>U2AF1 (NM_006758)</td>\n",
       "      <td>2 (15-41), 6 (137-161)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>XPO1 (NM_003400)</td>\n",
       "      <td>15 (563-575)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>146 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Gene                               Exons (codons) tested\n",
       "0     AKT1 (NM_005163)  3 (16-56), 4 (59-96), 6 (146-149), 6-7 (177-19...\n",
       "1     AKT2 (NM_001626)  3 (16-55), 4 (76-96), 6 (148-153), 7 (192-210)...\n",
       "2     AKT3 (NM_005465)  2 (16-28), 2 (48-58), 3 (66-95), 4 (123-143), ...\n",
       "3      ALK (NM_004304)                                  20-28 (1096-1388)\n",
       "4       AR (NM_000044)  1 (1-45), 1 (135-178), 1 (245-285), 1 (342-388...\n",
       "..                 ...                                                ...\n",
       "141   TP53 (NM_000546)                                       2-11 (1-394)\n",
       "142   TSC1 (NM_000368)                                      3-23 (1-1165)\n",
       "143   TSC2 (NM_000548)                   2-35 (1-1523), 37-42 (1555-1808)\n",
       "144  U2AF1 (NM_006758)                             2 (15-41), 6 (137-161)\n",
       "145   XPO1 (NM_003400)                                       15 (563-575)\n",
       "\n",
       "[146 rows x 2 columns]"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import dataset\n",
    "codons = pd.read_csv('STGA 2018 Codons Tested.csv')\n",
    "codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "c79a5d26-cb38-4326-b842-c0de4a3921b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene                     object\n",
       "Exons (codons) tested    object\n",
       "dtype: object"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codons.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "630932cd-5438-463f-b0d5-4cf2be04f507",
   "metadata": {},
   "source": [
    "## How to Do Regular Expressions in Python\n",
    "Source: https://docs.python.org/3/howto/regex.html\n",
    "- \"\\d\" matches any decimal digit, this is equivalent ot the class [0-9]\n",
    "- \"*\" specifies that the previous character can be matched zero or more times, instead of exactly once\n",
    "- \"+\" specifies that the previous character can be matched ONE or more times\n",
    "- \"?\" matches either once or zero times\n",
    "- \"\\section\" isolates text string to be matched. Regular expressions use the backslash character '\\' to indicate special forms or to allow special characters to be used without invoking their special meaning\n",
    "- \"findall()\" finds all substrings where the RE matches, and returns them as a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "e3ccbeef-5361-46d1-99b8-924ea80c4664",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Exons (codons) tested</th>\n",
       "      <th>Base Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AKT1 (NM_005163)</td>\n",
       "      <td>3 (16-56), 4 (59-96), 6 (146-149), 6-7 (177-19...</td>\n",
       "      <td>253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AKT2 (NM_001626)</td>\n",
       "      <td>3 (16-55), 4 (76-96), 6 (148-153), 7 (192-210)...</td>\n",
       "      <td>218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AKT3 (NM_005465)</td>\n",
       "      <td>2 (16-28), 2 (48-58), 3 (66-95), 4 (123-143), ...</td>\n",
       "      <td>195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ALK (NM_004304)</td>\n",
       "      <td>20-28 (1096-1388)</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AR (NM_000044)</td>\n",
       "      <td>1 (1-45), 1 (135-178), 1 (245-285), 1 (342-388...</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>TP53 (NM_000546)</td>\n",
       "      <td>2-11 (1-394)</td>\n",
       "      <td>394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>TSC1 (NM_000368)</td>\n",
       "      <td>3-23 (1-1165)</td>\n",
       "      <td>1165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>TSC2 (NM_000548)</td>\n",
       "      <td>2-35 (1-1523), 37-42 (1555-1808)</td>\n",
       "      <td>1777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>U2AF1 (NM_006758)</td>\n",
       "      <td>2 (15-41), 6 (137-161)</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>XPO1 (NM_003400)</td>\n",
       "      <td>15 (563-575)</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>146 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Gene                               Exons (codons) tested  \\\n",
       "0     AKT1 (NM_005163)  3 (16-56), 4 (59-96), 6 (146-149), 6-7 (177-19...   \n",
       "1     AKT2 (NM_001626)  3 (16-55), 4 (76-96), 6 (148-153), 7 (192-210)...   \n",
       "2     AKT3 (NM_005465)  2 (16-28), 2 (48-58), 3 (66-95), 4 (123-143), ...   \n",
       "3      ALK (NM_004304)                                  20-28 (1096-1388)   \n",
       "4       AR (NM_000044)  1 (1-45), 1 (135-178), 1 (245-285), 1 (342-388...   \n",
       "..                 ...                                                ...   \n",
       "141   TP53 (NM_000546)                                       2-11 (1-394)   \n",
       "142   TSC1 (NM_000368)                                      3-23 (1-1165)   \n",
       "143   TSC2 (NM_000548)                   2-35 (1-1523), 37-42 (1555-1808)   \n",
       "144  U2AF1 (NM_006758)                             2 (15-41), 6 (137-161)   \n",
       "145   XPO1 (NM_003400)                                       15 (563-575)   \n",
       "\n",
       "    Base Length  \n",
       "0           253  \n",
       "1           218  \n",
       "2           195  \n",
       "3           293  \n",
       "4           287  \n",
       "..          ...  \n",
       "141         394  \n",
       "142        1165  \n",
       "143        1777  \n",
       "144          52  \n",
       "145          13  \n",
       "\n",
       "[146 rows x 3 columns]"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Creating a new column for the base length of each gene\n",
    "codons['Base Length'] = None\n",
    "\n",
    "# Loops through each gene of the gene list, finds its base length, and adds it to the dataframe\n",
    "for index in range(codons.shape[0]):\n",
    "    # Converts the Exons of the selected gene to a string\n",
    "    current_exons = str(codons.iloc[index, 1])\n",
    "    \n",
    "    # Extracting start and end bases using regular expressions. \n",
    "    # Matches is a list of the substrings where the RE matches\n",
    "    matches = re.findall(r' \\((\\d+)-(\\d+)\\)', current_exons)\n",
    "    \n",
    "    # Calculating the length of each exon and summing them for selected gene\n",
    "    gene_base_length = 0\n",
    "\n",
    "    for match in matches:\n",
    "        start = int(match[0])  # Start base\n",
    "        end = int(match[1])  # End base\n",
    "\n",
    "        # Calculating the base length of each exon and adding it to the total base length for the selected gene\n",
    "        exon_base_length = end - start + 1\n",
    "        gene_base_length = gene_base_length + exon_base_length\n",
    "\n",
    "    # Setting the base length for the selected gene\n",
    "    codons.iloc[index,2] = gene_base_length\n",
    "    \n",
    "codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "beab1958-509c-48d3-8ddf-92d034d14f74",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total number of bases is 78669\n"
     ]
    }
   ],
   "source": [
    "# Calculating the total number of bases for the gene list \n",
    "num_bases = codons['Base Length'].sum()\n",
    "print('The total number of bases is ' + str(num_bases))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "848d810f-a992-475a-8db0-4d11d306671e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the DataFrame to an Excel file\n",
    "codons.to_excel('Total Number of Bases.xlsx', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16be8fd9-bc7e-414c-84ed-70797fac297f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
